velox/common/memory/SharedArbitrator.h

```diff
@@ -90,29 +92,70 @@ class SharedArbitrator : public memory::MemoryArbitrator {
   // The kind string of shared arbitrator.
   inline static const std::string kind_{"SHARED"};
 
+  struct ArbitrationOperation {
+    MemoryPool* const requestPool;
+    MemoryPool* const requestRoot;
+    const std::vector<std::shared_ptr<MemoryPool>>& candidatePools;
+    const uint64_t targetBytes;
+    const std::chrono::steady_clock::time_point startTime;
+
+    std::vector<Candidate> candidates;
+
+    bool needGlobalArbitration{false};
+    uint32_t numGlobalArbitrationAttempts{0};
+    uint64_t waitTimeUs{0};
+
+    explicit ArbitrationOperation(
+        uint64_t targetBytes,
+        const std::vector<std::shared_ptr<MemoryPool>>& candidatePools)
+        : ArbitrationOperation(nullptr, targetBytes, candidatePools) {}
+
+    ArbitrationOperation(
+        MemoryPool* _requestor,
+        uint64_t _targetBytes,
+        const std::vector<std::shared_ptr<MemoryPool>>& _candidatePools)
+        : requestPool(_requestor),
+          requestRoot(_requestor == nullptr ? nullptr : _requestor->root()),
+          candidatePools(_candidatePools),
+          targetBytes(_targetBytes),
+          startTime(std::chrono::steady_clock::now()) {
+      VELOX_CHECK(!candidatePools.empty());
+    }
+
+    void enterArbitration();
+    void leaveArbitration();
+  };
+
   class ScopedArbitration {
    public:
-    // Used by arbitration request NOT initiated from memory pool, e.g. through
-    // shrinkPools() API.
-    explicit ScopedArbitration(SharedArbitrator* arbitrator);
-
-    // Used by arbitration request initiated from a memory pool.
-    explicit ScopedArbitration(
-        MemoryPool* requestor,
-        SharedArbitrator* arbitrator);
+    // Used by arbitration request initiated from a memory pool or memory
+    // capacity shrink request sent through shrinkPools() API.
+    ScopedArbitration(SharedArbitrator* arbitrator, ArbitrationOperation* op);
 
     ~ScopedArbitration();
 
    private:
-    MemoryPool* const requestor_;
+    ArbitrationOperation* const operation_;
     SharedArbitrator* const arbitrator_;
-    const std::chrono::steady_clock::time_point startTime_;
     const ScopedMemoryArbitrationContext arbitrationCtx_;
+    const std::chrono::steady_clock::time_point startTime_;
+  };
+
+  struct ArbitrationState {
+    ArbitrationOperation* current;
+
+    // The promises of the arbitration requests waiting for the serialized
```

Comment:
The promises of the arbitration requests waiting for the serially executed arbitration.